xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -0,0 +1,725 @@
+   * never care about others.
+   */
+  @Nullable
+  private String processRouteConfig(RouteConfiguration config) {
+    List<VirtualHost> virtualHosts = config.getVirtualHostsList();
+    String clusterName = null;
+    // Proceed with the virtual host that has longest wildcard matched domain name with the
+    // hostname in original "xds:" URI.
+    int matchingLen = -1;  // longest length of wildcard pattern that matches host name
+    for (VirtualHost vHost : virtualHosts) {
+      for (String domain : vHost.getDomainsList()) {
+        // We only validate data for VirtualHosts that matches hostname.
+        if (matchHostName(hostName, domain) && domain.length() > matchingLen) {
+          matchingLen = domain.length();
+          // The client will look only at the last route in the list (the default route),
+          // whose match field must be empty and whose route field must be set.
+          List<Route> routes = vHost.getRoutesList();
+          if (!routes.isEmpty()) {
+            Route route = routes.get(routes.size() - 1);
+            // TODO(chengyuanzhang): check the match field must be empty.
+            if (route.hasRoute()) {
+              clusterName = route.getRoute().getCluster();
+            }
+          }
+        }
+      }
+    }
+    return clusterName;
+  }
+
+  @VisibleForTesting
+  final class RpcRetryTask implements Runnable {
+    @Override
+    public void run() {
+      startRpcStream();
+      if (configWatcher != null) {
+        adsStream.sendXdsRequest(ADS_TYPE_URL_LDS, ImmutableList.of(ldsResourceName));
+      }
+      // TODO(chengyuanzhang): send CDS/EDS requests if CDS/EDS watcher presents.
+    }
+  }
+
+  private final class AdsStream implements StreamObserver<DiscoveryResponse> {
+    private final AggregatedDiscoveryServiceGrpc.AggregatedDiscoveryServiceStub stub;
+
+    private StreamObserver<DiscoveryRequest> requestWriter;
+    private boolean responseReceived;
+    private boolean closed;
+
+    // Last successfully applied version_info for each resource type. Starts with empty string.
+    // A version_info is used to update management server with client's most recent knowledge of
+    // resources.
+    private String ldsVersion = "";
+    private String rdsVersion = "";
+
+    // Response nonce for the most recently received discovery responses of each resource type.
+    // Client initiated requests start response nonce with empty string.
+    // A nonce is used to indicate the specific DiscoveryResponse each DiscoveryRequest
+    // corresponds to.
+    // A nonce becomes stale following a newer nonce being presented to the client in a
+    // DiscoveryResponse.
+    private String ldsRespNonce = "";
+    private String rdsRespNonce = "";
+
+    // Most recently requested resource name(s) for each resource type. Note the resource_name in
+    // LDS requests will always be "xds:" URI (including port suffix if present).
+    @Nullable
+    private String rdsResourceName;
```

Comment:
Any reason why this is a member of `AdsStream` but `ldsResourceName` is a member of `XdsClientImpl` . There is 1-to-1 relationship between `AdsStream` and `XdsClientImpl` so I don't know why they are separate.

Reply:
`ldsResourceName` never changes once it is set for an `XdsClient` instance (maybe giving it a different name would make this clearer). `rdsResourceName` is the intermediate resource name extracted from a received LDS response and to be used in RDS requests. It may change if receiving new LDS responses. As the comment says, it is the most recently requested resource name, which is `AdsStream`'s state.